File: computing/interview/design-patterns/structural/decorator/csharp/DecoratorDemo.cs
using System;

namespace Interview.DesignPatterns.Structural.Decorator
{
    public interface IDataSource
    {
        string Read();
    }

    public sealed class FileDataSource : IDataSource
    {
        private readonly string _contents;

        public FileDataSource(string contents)
        {
            _contents = contents;
        }

        public string Read() => _contents;
    }

    public abstract class DataSourceDecorator : IDataSource
    {
        protected DataSourceDecorator(IDataSource wrappee)
        {
            Wrappee = wrappee;
        }

        protected IDataSource Wrappee { get; }
        public abstract string Read();
    }

    public sealed class EncryptionDecorator : DataSourceDecorator
    {
        public EncryptionDecorator(IDataSource wrappee) : base(wrappee)
        {
        }

        public override string Read() => $"<encrypted>{Wrappee.Read()}</encrypted>";
    }

    public sealed class CompressionDecorator : DataSourceDecorator
    {
        public CompressionDecorator(IDataSource wrappee) : base(wrappee)
        {
        }

        public override string Read() => $"<compressed>{Wrappee.Read()}</compressed>";
    }

    public static class Demo
    {
        public static void Main()
        {
            IDataSource source = new FileDataSource("salaries.csv");
            IDataSource encrypted = new EncryptionDecorator(source);
            IDataSource secured = new CompressionDecorator(encrypted);
            Console.WriteLine(secured.Read());
        }
    }
}

File: computing/interview/design-patterns/structural/decorator/java/DecoratorDemo.java
package interview.designpatterns.structural.decorator;

interface DataSource {
    String read();
}

final class FileDataSource implements DataSource {
    private final String contents;

    FileDataSource(String contents) {
        this.contents = contents;
    }

    @Override
    public String read() {
        return contents;
    }
}

abstract class DataSourceDecorator implements DataSource {
    private final DataSource wrappee;

    protected DataSourceDecorator(DataSource wrappee) {
        this.wrappee = wrappee;
    }

    protected DataSource wrappee() {
        return wrappee;
    }
}

final class EncryptionDecorator extends DataSourceDecorator {
    EncryptionDecorator(DataSource wrappee) {
        super(wrappee);
    }

    @Override
    public String read() {
        return "<encrypted>" + wrappee().read() + "</encrypted>";
    }
}

final class CompressionDecorator extends DataSourceDecorator {
    CompressionDecorator(DataSource wrappee) {
        super(wrappee);
    }

    @Override
    public String read() {
        return "<compressed>" + wrappee().read() + "</compressed>";
    }
}

public final class DecoratorDemo {
    private DecoratorDemo() {}

    public static void main(String[] args) {
        DataSource source = new FileDataSource("salaries.csv");
        DataSource secured = new CompressionDecorator(new EncryptionDecorator(source));
        System.out.println(secured.read());
    }
}
